learns-app-content/week-08/discussions.md

```diff
@@ -1,26 +1,319 @@
 
-#### Topic n Check-for-understanding Questions
+function FilteredList({ items }) {
+  const [query, setQuery] = useState('');
 
-### Review and Summary
+  const filtered = useMemo(() => {
+    console.log('Filtering items...');
+    return items.filter((item) =>
+      item.toLowerCase().includes(query.toLowerCase()),
+    );
+  }, [items, query]);
 
-- Key takeaways from this lesson  
-- Connection to next lesson
-- Quick preview of this week’s coding assignment
+  return (
+    <div>
+      <input
+        placeholder="Search..."
+        value={query}
+        onChange={(e) => setQuery(e.target.value)}
+      />
+      <ul>
+        {filtered.map((item) => (
+          <li key={item}>{item}</li>
+        ))}
+      </ul>
+    </div>
+  );
+}
+```
+
+Every time the user types, React re-renders the component — but `useMemo` ensures the filtering logic only runs when `items` or `query` actually change. Without it, even unrelated state updates would trigger a full re-filtering of the list.
+
+**Key Points about useMemo:**
+
+- Returns a **memoized value** (the result of the computation)
+- Only recalculates when dependencies change
+- Useful for expensive calculations that shouldn't run on every render
+- The function inside cannot accept parameters — it uses values from the surrounding scope
+
+#### useCallback: Caching the Function Definition Itself
+
+While `useMemo` caches a result, `useCallback` caches a function.
+
+Imagine you're passing a callback to a child component that re-renders whenever that callback changes. Since React redefines all functions inside a component body on every render, that child would re-render unnecessarily.
+
+We can fix that using `useCallback`:
+
+```jsx
+import { useCallback, useState } from 'react';
+
+function CounterButton({ onIncrement }) {
+  console.log('Button rendered');
+  return <button onClick={onIncrement}>Increment</button>;
+}
+
+export default function Counter() {
+  const [count, setCount] = useState(0);
+
+  const handleIncrement = useCallback(() => {
+    setCount((c) => c + 1);
+  }, []); // function only recreated if dependencies change
+
+  return (
+    <div>
+      <p>Count: {count}</p>
+      <CounterButton onIncrement={handleIncrement} />
+    </div>
+  );
+}
+```
+
+Here, `useCallback` ensures `handleIncrement` maintains the same reference between renders, so the `CounterButton` component doesn't re-render unnecessarily.
+
+**Key Points about useCallback:**
+
+- Returns a **memoized function** (not the result of calling it)
+- Prevents function recreation on every render
+- Essential when passing callbacks to optimized child components
+- Uses the setter function pattern to avoid stale closures
+
+### Optimizing React
+
+#### Compare and Contrast: useMemo vs useCallback
+
+Now that we understand both hooks individually, let's explore how they relate to each other. Here's a crucial insight: `useCallback` is actually just a specialized version of `useMemo`!
```

Comment:
Saying that `useCallback` is a specialized version of `useMemo` will introduce confusion. Even if we provide technical details of how React implements these hooks, it would detract from the flow and clarity of this section.